protocols/purple/purple.c

```diff
@@ -1055,9 +1076,50 @@ static void prplcb_request_test()
 }
 */
 
+void* prplcb_request_input(const char *title, const char *primary,
+        const char *secondary, const char *default_value, gboolean multiline,
+        gboolean masked, gchar *hint, const char *ok_text, GCallback ok_cb,
+        const char *cancel_text, GCallback cancel_cb, PurpleAccount *account,
+        const char *who, PurpleConversation *conv, void *user_data)
+{
+	struct im_connection *ic = purple_ic_by_pa(account);
+	struct purple_data *pd = ic->proto_data;
+	guint id = pd->next_request_id;
+	struct request_input_data *ri = g_new0(struct request_input_data, 1);
+
+	pd->next_request_id = (id + 1) % 10000;
+	ri->data_callback = (ri_callback_t) ok_cb;
+	ri->user_data = user_data;
+	g_hash_table_insert(pd->input_requests, GUINT_TO_POINTER(id), ri);
+
+	char buddy[20] = { 0 };
+	sprintf(buddy, "%s_%d", PURPLE_REQUEST_HANDLE, id);
```

Comment:
You should probably just `g_strdup_printf()`.  Yes, it is "inefficient" by not working right on the stack, but only by a slight amount.  This will also allow for the full range of `guint` rather than just `0 - 9999`, granted that limit would probably never be reached.  Also, you should be using `%u`, not `%d`.